fe/fe-core/src/main/java/org/apache/doris/resource/resourcegroup/QueryQueue.java

```diff
@@ -0,0 +1,103 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.resource.resourcegroup;
+
+import com.google.common.base.Preconditions;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+// note(wb) refer java BlockingQueue, but support altering capacity
+// todo(wb) add wait time to profile
+public class QueryQueue {
+
+    private static final Logger LOG = LogManager.getLogger(QueryQueue.class);
+    // note(wb) use fair strategy for FIFO
+    // but maybe display lower overall throughput
+    // need performance test later
+    private final ReentrantLock queueLock = new ReentrantLock(true);
+    private final Condition queueLockCond = queueLock.newCondition();
+    // resource group property
+    private int maxRunningQueryNum;
+    private int maxWaitingQueryNum;
+    private int queryWaitTimeout; // ms
+    // running property
+    private int currentRunningQueryNum;
+    private int currentWaitingQueryNum;
+
+    public QueryQueue(int maxRunningQueryNum, int maxWaitingQueryNum, int queryWaitTimeout) {
+        this.maxRunningQueryNum = maxRunningQueryNum;
+        this.maxWaitingQueryNum = maxWaitingQueryNum;
+        this.queryWaitTimeout = queryWaitTimeout;
+    }
+
+    public void offer() throws InterruptedException {
+        // to prevent hang
+        // the lock shouldn't be hold for too long
+        // we should catch the case when it happens
+        queueLock.tryLock(5, TimeUnit.SECONDS);
+        try {
+            // currentRunningQueryNum may bigger than maxRunningQueryNum
+            // because maxRunningQueryNum can be altered
+            if (currentRunningQueryNum >= maxRunningQueryNum) {
+                if (currentWaitingQueryNum >= maxWaitingQueryNum) {
+                    throw new QueryQueueException(
+                            "query waiting queue is full, queue length=" + maxWaitingQueryNum);
+                }
+
+                currentWaitingQueryNum++;
+                boolean ret = queueLockCond.await(queryWaitTimeout, TimeUnit.MILLISECONDS);
+                currentWaitingQueryNum--;
+                if (!ret) {
+                    throw new QueryQueueException("query wait timeout " + queryWaitTimeout + " ms");
+                }
+            }
+            currentRunningQueryNum++;
+        } finally {
+            queueLock.unlock();
+        }
+    }
+
+    public void poll() throws InterruptedException {
+        queueLock.tryLock(5, TimeUnit.SECONDS);
+        try {
+            currentRunningQueryNum--;
```

Comment:
Whether to print the log fe.log regularly, print `currentRunningQueryNum` and `currentWaitingQueryNum`

Reply:
Good idea. Currently we can print it when exception happends.